### DISCUSSION TITLE
38 - Links to non-note files

### DISCUSSION THREAD

**Discussion by SeriousBug** [dendronhq/dendron#1975] [RFCs/Ideas]
This is a discussion for RFC 38, "Links to non-note files".

https://docs.dendron.so/notes/WCZ1LzhM07Om3KFPERk6l.html

**Comment by kevinslin** [comment]
for clarification, the `remote` field in the frontmatter for proxy notes is to have a remote fallback in case the local version is un-availble. 

the use case: for files like images and videos, I might not want to keep them in my local folder but have them available via a CDN. a CDN makes it so that I can access those files anywhere (we can also make it private by adding a auth proxy with token authentication). Dendron's local first principle means that we still want to provide a local option. So when we access files from the CDN, dendron should also cache a copy locally. 

As a simplication, if the file is located remotely, we can avoid hardcoding a local url and generate the url automatically via convention (eg. `assets/{hash}.{extension}`)

That way, the proxy might look like this


```yml
remote: true
url: cloudfront.com/1234.jpg
# local url is generated automatically
```

**Comment by SeriousBug** [comment]
Related: #1984

**Reply by ScriptAutomate** [reply]
- Related, when it comes to image files: https://github.com/dendronhq/dendron/issues/1450
- Related, to things like proxy notes, when it comes to embedding things like scripts: https://github.com/dendronhq/dendron/issues/1565

**Comment by nickolay-kondratyev** [comment]
Something we have discussed previously, if the use case is to select particular lines of source code to reference this
  ```
  [[src/example.cpp#L12:#L16]]
  ```
This has high likelehood to keep getting out of date very quickly as code changes. 
(Unless the code base is written out of tiny files that very very rarely change).

**Reply by kevinslin** [reply]
for line items, we'll probably end up relying on anchors which will let us select arbitrary ranges of text and will work on any file that supports comments: `[[src/example.cpp#^start:#^end]]`

**Reply by ScriptAutomate** [reply]
I like the idea of anchors that Kevin mentions above, because then you can use things like comments inside of code files that could represent start/end of sections.

**Reply by jonathanyeung** [reply]
But how does this scale to all file types? I'm not too keen on the concept of adding block anchors to external files.  What if those files can't be modified? What if the target document type has syntax that interferes with our block anchor syntax?

**Reply by kevinslin** [reply]
its not meant for all file types, but all file types that support comments (aka code)

**Comment by nickolay-kondratyev** [comment]
> For example, the MIME type for javascript code is `text/javascript`, which Dendron could use to set highlighting for javascript.

We should also consider auto detecting the common file extensions if MIME type is not specified. js->javascript ts->typescript etc...

**Comment by SR--** [comment]
For images and video files that come with standard metadata, might it be possible to populate proxy notes with relevant content, for example at the `paste` stage?

1. Use `copy image` on a web site
2. Extract metadata from the image at `paste image` [cf. IPTC Core](https://iptc.org/standards/photo-metadata/iptc-standard/)
3. Enter this data into schema-driven fields within the proxy note.

Allow users to select what fields are in use.

4. Make all fileds in frontmatter and in the proxy note selectable for display at embed time, in another note and also in the graph.

This information can provide credits, dates, places, image captions, etc.

- Potentially allow for this information to be baked back into image files.

- Support refactoring tools for mass editing.

- Support searching/quering of the metadata.

- Support embedding based on queries.

**Comment by SR--** [comment]
Will proxy notes have one-to-one relationship with non-note files?

I would love to have a way to create draft compositions and aggregations of arbitrary files, notes and not, for creative development. A pin board is a useful metaphor, or sticky notes. Note sure if a proxy note with multiple non-note files in it might be a vehicle for this.

**Reply by SeriousBug** [reply]
Proxy notes likely should be one-to-one for simplicity, but would collecting proxy notes under a hierarchy achieve what you're thinking of? You could even use a wildcard reference to embed them all into a note like `![[proxy,inspirations,*]]`. Did you have any other features in mind?

**Reply by SR--** [reply]
I am thinking of the flexibility intended by the tools such as [Scrivener](https://www.literatureandlatte.com/scrivener/overview) and the many ways one can approch the information there. More or less detailed. Arranged in different types of layouts, etc. Dendron can do some much better than Scrivener given its existing foundation.

![2021-10-11-21-04-45](https://user-images.githubusercontent.com/1008059/147653973-f42f0b2b-684d-4f05-a68c-96a0efcd9a49.png)

A note has only one place in the hierarchy. Embedding makes a step in the right direction but it is not sufficiently flexible.

I fancy a way to have multiple refactorable hierachies or orders for individual projects and drafts within projects without the need to change the master hierarchy that organises the overall knowledge system.

**Comment by SR--** [comment]
- Ability to treat files similar to notes might include
  - Ability to rename files with their references via a Rename Note command (already mentioned, I think but not sure of the exact capabilities)
  - Ability to delete files with their references. And ability to delete a note with all of its referenced (with a preview showing dependences in other notes/graph views, etc.)
  - Backlinks to files (proxy notes will work but maybe not in all scenarious?)
  - Graphing of files, this can be another type of link in addition to the two we have now: hierarchy and reference; or to be considered (again, will depend on what proxy notes do)
- Allow finding notes with files based on parameters such as: number, size of files, types of files, file metadata.
- Synchronisation between Preview, file view, graph and markdown for easier editing and finding of files. Selecting a file/image in any view (file list, markdown, preview) should select the file/image in all views and put it into focus in the UI

**Comment by jonathanyeung** [comment]
I think if we do decide to do proxy notes, we should have it be a part of the note trait system.

**Comment by bitdryvr** [comment]
I am not a professional developer, but can usually read and understand code from others. My talents are as a software architect, software tester, and overall technology consultant. As such, I do not see myself authoring the original file but adding important comments back to the author.

How can I best get involved to understand this better and contribute?

**Reply by SeriousBug** [reply]
You're always welcome to give feedback on our discussions and send feature requests! Another great contribution is updating our documentation (dendron-site repository) to make it easier to understand, and to update anything out of date.

For this project specifically, it's in the design stage while we focus on making Dendron easier to use. In the meantime, you could update the RFC linked at the top (fork the repository, make changes, then send a pull request). It would be great if someone put together all the community feedback into features we can work on and put them in the RFC.

**Reply by bitdryvr** [reply]
> It would be great if someone put together all the community feedback into features we can work on and put them in the RFC.

Are you specifically meaning the comments here on this github discussion page or is there other feedback to consider?

**Reply by SeriousBug** [reply]
Yes this page.

**Reply by bitdryvr** [reply]
It would seem easy to archive the rich text as a PDF, but that is hard to use later. Were you thinking of something that preserves the rich text into a simple MD note? Are there any examples that this should be similar to?

**Reply by bitdryvr** [reply]
I have tried both a PDF archive and a GDOC copy-paste. The PDF seems better to me as it is a complete rendered snapshot that preserves all of the rich formattings of both Github and the authors, including conversation threading, user names, avatars, images, highlighting, quotes, code blocks, etc.

Any copy-paste will potentially lose some of this meaning. The person performing the copy-paste archive may interpret things incorrectly and change the meaning.

I am inclined to just capture a PDF snapshot unless someone has a better idea.

Thoughts?

**Reply by bitdryvr** [reply]
I have heard nothing to dissuade me away from PDF. It seems a good way to capture this rich content while preserving both the text and other elements for future use.

Both Chrome and Firefox make it very easy to print to PDF at any point. So it makes sense to capture it just after the discussion is closed. Do you know when that will be?

**Reply by SeriousBug** [reply]
> Both Chrome and Firefox make it very easy to print to PDF at any point. So it makes sense to capture it just after the discussion is closed. Do you know when that will be?

I'm not exactly sure what you are asking.

- If you are talking about capturing this discussion page and adding it to Dendron, that's not what I asked when I said you could contribute. I meant summarizing the discussion here and boiling it down to features that I can then add to Dendron. We don't really need to capture the whole discussion, it will always be up here on Github for reference.
- If you are talking about your own Dendron workspace, and adding PDFs into that workspace, you can do that by adding the PDF files into a vaults `assets` folder.

**Reply by bitdryvr** [reply]
Ignore my question as I was mistakenly thinking about how to capture this discussion for archive purposes and was concentrating on when to perform this snapshot. As you point out it does not matter as the conversation is not going away.

**Comment by bitdryvr** [comment]
Is there any code to examine or test? I just cloned the dendron repo so I think I am ready.

**Reply by SeriousBug** [reply]
Cool, here is the development docs for reference: https://docs.dendron.so/notes/3489b652-cd0e-4ac8-a734-08094dc043eb/

I'm not sure if there's anything specific we need tested, but if there are any parts of Dendron where you think you experienced problems those would be great to look at. You can do some manual testing and let us know if you hit any bugs, or write automated tests and send a pull request for it (there are existing tests in the repo, and docs should have more information).

**Comment by bitdryvr** [comment]
@kevins8 and I have been having a parallel conversation that turns out that is related to this topic...

- Dendron can provide value even if you don't own the files or cannot make changes to the source (non-note) files. Dendron proxy notes can provide an intermediate solution to link to dendron notes to the source files.

- It is very strong for coders and development teams where you have write access (and permission) to the source, but I envision cases where I have a local copy of something I have found. The local copy is an offline copy for archiving. I also envision pointing to completely external sources. Both of these categories would benefit from a dendron comment layer for my thoughts.

- Dendron already has a layered architecture (https://docs.dendron.so/notes/74UyCLyLvCxFTQNgGgSRW.html) that works well for VSCode originals, but if we can swap this layer for other document types then we can extend dendron. 

- I use notes mainly for personal knowledge management, but with an eye on nuggets of information to highlight to others. I find myself curating sources (podcasts, tools, websites, etc.), adding my own comments, and then sharing them with others as needed. I want to organize **both** the source material and my comments. Some of the source materials cannot be easily downgraded to simple text. Some (like PDFs) may, but others (like YouTube videos), not so much.

- Linking dendron to the source files can be thought of as adding a dendron comment layer to the base layer of the original document.

- Future extensions to dendron can be made to add other source file types to the supported list.

**Reply by bitdryvr** [reply]
Sorry, mentioned the wrong user ID as that is the handle on discord. It should be @kevinslin here.

**Comment by SeriousBug** [comment]
> Makes sense to implement this behavior as a note trait that is used by proxy notes. It should be a static property mapped to an enum. By default, value is `fname` (open note using file name). Can be changed to a different target (eg. frontmatter property). Should use same syntax as [[44 Standard srcfieldMapping for all pods|dendron://dendron.docs/rfc.44-standard-src-field-Mapping-for-all-pods]]
>
> For external resource, we can specify a trait to open by `url`. Dendron has an `OpenLink` command to open web urls. We can incorporate the `open link` functionality in here so that "GoToNote` can still open the underlying note

**Reply by robotcorner** [reply]
What is the progress on this? If I'm not mistaken you tagged yourself as the implementer of this, but I haven't seen any update in a while. It's been in implementation phase for a few months with no updates. This is the biggest thing keeping me out of dendron. I have too many notes I want to import that I don't want to convert to markdown. I have years worth of .docx, .pdf, powerpoint files, libre office files, etc.

**Reply by kevinslin** [reply]
Hey @robotcorner - this feature is now partially implemented via the `Go to` command which can navigate directly to a URI -> https://wiki.dendron.so/notes/vikda112zw2sss8lbit4qey

Currently the `uri` parameter only accepts web urls. The next step would be to make the `uri` parameter also accept local files.

**Reply by dionysuzx** [reply]
that's awesome news! i just heard about this thread. i definitely would like this as well for locally linking pdfs

**Comment by serverhorror** [comment]
I think #533 is also related to this.

As for myself, I'd be very interested in getting:

* GitHub
* Gitlab
* BitBucket
* generally "raw" content with code highlighting

code urls to work (hopefully allowing something like `.../source.go#L9-L15`).
